Single Use Links UI

awead opened the pull request:
Fixes #278 

Here's the preliminary idea.

![sul](https://cloud.githubusercontent.com/assets/312085/17064086/b563a43e-5008-11e6-814b-011f036ea91b.png)

Need some feedback at this point before it can be finished.

@projecthydra/sufia-code-reviewers

mjgiarlo commented:
The last few times we discussed this, I believe we concluded that show links were more trouble than they were worth because they display a site full of links, only some of which will work for the user. It's an awkward user experience. Before I go over the changes in full, I'd lean towards excising that change but take that with a grain of salt. I'll review this fully later today when I'm done with my meetings.

mjgiarlo commented on app/controllers/concerns/curation_concerns/single_use_links_controller_behavior.rb:
```diff
@@ -25,6 +25,16 @@ def create_show
     end
 
+    def index
+      links = SingleUseLink.where(itemId: params[:id])
```
`s/itemId/item_id/`

mjgiarlo commented on app/controllers/concerns/curation_concerns/single_use_links_controller_behavior.rb:
```diff
@@ -25,6 +25,16 @@ def create_show
 
+    def index
+      links = SingleUseLink.where(itemId: params[:id])
+      render partial: 'curation_concerns/file_sets/single_use_link_rows', locals: { single_use_links: links }
```
I believe this can be simplified to:

```ruby
render 'curation_concerns/file_sets/single_use_link_rows', single_use_links: links
```

awead replied:
@mjgiarlo you get `ActionView::MissingTemplate - Missing template curation_concerns/file_sets/single_use_link_rows` because it's expecting it to be a view template and not a partial since we'er in the `index` action.

mjgiarlo replied:
I stand corrected. Context is everything. :) Thanks, @awead.

mjgiarlo commented on app/presenters/curation_concerns/file_set_presenter.rb:
```diff
@@ -33,5 +33,9 @@ def page_title
     end
+
+    def single_use_links
+      @single_use_links ||= SingleUseLink.where(itemId: id)
```
`s/itemId/item_id/`

mjgiarlo commented on app/views/curation_concerns/file_sets/_single_use_link_rows.html.erb:
```diff
@@ -0,0 +1,18 @@
+  <% url_helper = link.path =~ /downloads/ ? 'download_single_use_link_url' : 'show_single_use_link_url' %>
+  <tr>
+    <td><%= type %></td>
+    <td><%= link.downloadKey %></td>
```
I'm `+0` on displaying the download key to users. Not sure it's meaningful for many of them. Certainly doesn't *hurt* to show it, so I won't complain too much about this.

awead replied:
It exposes everything, but the distinction isn't really explained. If we only want to show one kind of link, I would argue that it should be the "show" link, although we don't have to name it as such. In the process of visiting a show link, a download link is created automatically. However, if the user visited the show link, but never actually downloaded the content, the auto-created download link would still be hanging around. The user might want to be able to delete it, if they felt the need to control access, but no one would know the key of that download link anyway, so it's probably a moot issue.

mjgiarlo commented on app/views/curation_concerns/file_sets/_single_use_links.html.erb:
```diff
@@ -0,0 +1,20 @@
+  </thead>
+  <tbody data-url="<%= curation_concerns.generated_single_use_links_path(presenter) %>">
+    <% if presenter.single_use_links.empty? %>
+      <tr><td>No links have been genereated</td></tr>
```
`s/genereated/generated/`

mjgiarlo commented on app/views/curation_concerns/file_sets/_single_use_links.html.erb:
```diff
@@ -0,0 +1,20 @@
+    <% if presenter.single_use_links.empty? %>
+      <tr><td>No links have been genereated</td></tr>
+    <% else %>
+      <%= render partial: 'single_use_link_rows', locals: { single_use_links: presenter.single_use_links } %>
```
Can be simplified to:

```ruby
render 'single_use_link_rows', single_use_links: presenter.single_use_links 
```

awead replied:
this works.

mjgiarlo commented on spec/factories/single_use_links.rb:
```diff
@@ -0,0 +1,13 @@
+FactoryGirl.define do
+  factory :single_use_link do
+    factory :show_link do
+      itemId 'fs-id'
```
`s/itemId/item_id/`

mjgiarlo commented on spec/factories/single_use_links.rb:
```diff
@@ -0,0 +1,13 @@
+    end
+
+    factory :download_link do
+      itemId 'fs-id'
```
`s/itemId/item_id/`

mjgiarlo commented on spec/presenters/curation_concerns/file_set_presenter_spec.rb:
```diff
@@ -77,4 +77,11 @@
+
+  describe "single_use_links" do
+    it "returns ActiveRecord::Relation of all single use links for the file set" do
+      expect(SingleUseLink).to receive(:where).with(itemId: presenter.id)
```
`s/itemId/item_id/`

awead replied:
`itemId` is coming from the model https://github.com/projecthydra/curation_concerns/blob/master/app/models/single_use_link.rb#L8

mjgiarlo replied:
Ah, yes, thanks. I'll retract those comments.

awead commented:
@mjgiarlo I've addressed your comments, and have redone the JS in ECMA6 form, with some tests. It's not super-covered, but it's better than the previous non-existent coverage!

mjgiarlo commented:
As soon as Travis is :green_apple:, I'll be happy to merge it, show links and all.

awead commented:
@mjgiarlo this is ready. I added one more commit for engine cart, but it wasn't impacting the build.

awead commented:
@mjgiarlo this needs some more work to get the Js code working with Sufia.

mjgiarlo commented:
@awead OK. Thanks for testing Sufia integration before I merged this.

awead commented:
@mjgiarlo this is ready to go now, and I have a follow-up PR in Sufia that makes use of these changes.

mjgiarlo commented:
👏
